File: android/src/main/java/uk/co/workingedge/gemini/x/GeminiXPlugin.kt
package uk.co.workingedge.gemini.x

import android.graphics.Bitmap
import android.util.Base64

import com.getcapacitor.JSObject
import com.getcapacitor.Plugin
import com.getcapacitor.PluginCall
import com.getcapacitor.PluginMethod
import com.getcapacitor.annotation.CapacitorPlugin

import org.json.JSONArray
import org.json.JSONObject

import uk.co.workingedge.gemini.x.lib.BlobHistoryPart
import uk.co.workingedge.gemini.x.lib.GeminiX
import uk.co.workingedge.gemini.x.lib.HistoryItem
import uk.co.workingedge.gemini.x.lib.HistoryPart
import uk.co.workingedge.gemini.x.lib.Image
import uk.co.workingedge.gemini.x.lib.ImageHistoryPart
import uk.co.workingedge.gemini.x.lib.TextHistoryPart

@CapacitorPlugin(name = "GeminiX")
class GeminiXPlugin : Plugin() {
    private val GeminiXResponseChunkEvent = "GeminiXResponseChunk"
    /*************************************************************************
     * Plugin Methods
     ************************************************************************/

    @PluginMethod
    fun initModel(call: PluginCall) {
        try {
            val params = call.getObject("params")

            // Required params
            val modelName = params.getString("modelName")!!
            val apiKey = params.getString("apiKey")!!

            // Optional params
            var temperature: Float? = null
            if (params.has("temperature")) {
                temperature = params.getLong("temperature").toFloat()
            }

            var topK: Int? = null
            if (params.has("topK")) {
                topK = params.getInt("topK")
            }

            var topP: Float?  = null
            if (params.has("topP")) {
                topP = params.getLong("topP").toFloat()
            }

            var maxOutputTokens: Int?  = null
            if (params.has("maxOutputTokens")) {
                maxOutputTokens = params.getInt("maxOutputTokens")
            }

            var stopSequences: List<String>? = null
            if (params.has("stopSequences")) {
                val stopSequenesJSONArray = params.getJSONArray("stopSequences")
                val stopSequencesArray = arrayListOf<String>()
                for (i in 0 until stopSequenesJSONArray.length()) {
                    stopSequencesArray.add(stopSequenesJSONArray.getString(i))
                }
                stopSequences = stopSequencesArray.toList()
            }

            val config = mutableMapOf<String, Any>()
            if (temperature != null) {
                config["temperature"] = temperature
            }
            if (topK != null) {
                config["topK"] = topK
            }
            if (topP != null) {
                config["topP"] = topP
            }
            if (maxOutputTokens != null) {
                config["maxOutputTokens"] = maxOutputTokens
            }
            if (stopSequences != null) {
                config["stopSequences"] = stopSequences
            }

            var safetySettings: Map<String, String>? = null
            if (params.has("safetySettings")) {
                val safetySettingsJSONObject = params.getJSONObject("safetySettings")
                val safetySettingsMap = mutableMapOf<String, String>()
                for (key in safetySettingsJSONObject.keys()) {
                    safetySettingsMap[key] = safetySettingsJSONObject.getString(key)
                }
                safetySettings = safetySettingsMap.toMap()
            }

            GeminiX.init(modelName, apiKey, config, safetySettings)

            call.resolve()
        } catch (e: Exception) {
            call.reject(e.message)
        }
    }

    @PluginMethod
    fun sendMessage(call: PluginCall) {
        try {
            var streamResponse = false
            var images = JSONArray()

            val inputText = call.getString("inputText")!!
            if(call.hasOption("options")){
                val opts = call.getObject("options")
                if(opts.has("streamResponse")){
                    streamResponse = opts.getBoolean("streamResponse")
                }
                if(opts.has("images")){
                    images = opts.getJSONArray("images")
                }
            }

            val modelImages:List<Image> = GeminiX.getModelImages(images, this.context)

            GeminiX.sendMessage(
                { response, isComplete ->
                    val result = JSObject()
                    result.put("response", response)
                    if(isComplete){
                        call.resolve(result)
                    }else{
                        result.put("isChat", false)
                        bridge.triggerWindowJSEvent(GeminiXResponseChunkEvent, result.toString())
                    }
                },
                { error ->
                    call.reject(error)
                }, inputText, modelImages, streamResponse)
        } catch (e: Exception) {
            call.reject(e.message)
        }
    }

    @PluginMethod
    fun countTokens(call: PluginCall) {
        try {
            var images = JSONArray()

            val inputText = call.getString("inputText")!!
            if(call.hasOption("options")){
                val opts = call.getObject("options")
                if(opts.has("images")){
                    images = opts.getJSONArray("images")
                }
            }

            val modelImages:List<Image> = GeminiX.getModelImages(images, this.context)

            GeminiX.countTokens(
                { response ->
                    val result = JSObject()
                    result.put("count", response)
                    result.put("isChat", false)
                    call.resolve(result)
                },
                { error ->
                    call.reject(error)
                }, inputText, modelImages)
        } catch (e: Exception) {
            call.reject(e.message)
        }
    }

    @PluginMethod
    fun initChat(call: PluginCall) {
        try {
            val history: MutableList<HistoryItem> = mutableListOf()
            if(call.hasOption("chatHistory")){
                val jsonHistory = call.getArray("chatHistory")
                for (i in 0 until jsonHistory.length()) {
                    val item = jsonHistory.getJSONObject(i)
                    val isUser = item.getBoolean("isUser")

                    val historyParts: MutableList<HistoryPart> = mutableListOf()

                    if(item.has("text")){
                        val text = item.getString("text");
                        val historyPart:HistoryPart = TextHistoryPart(text)
                        historyParts.add(historyPart)
                    }

                    if(item.has("images")){
                        val images = item.getJSONArray("images")
                        for (j in 0 until images.length()) {
                            val image = images.getJSONObject(j)
                            val uri = image.getString("uri")
                            val bitmap = GeminiX.getBitmapFromUri(uri, this.context)
                            var historyPart:HistoryPart?
                            if(image.has("mimeType")){
                                val mimeType = image.getString("mimeType")
                                val blob = GeminiX.bitmapToByteArray(bitmap)
                                historyPart = BlobHistoryPart(blob, mimeType)
                            }else{
                                historyPart = ImageHistoryPart(bitmap)
                            }
                            historyParts.add(historyPart)
                        }
                    }
                    val historyItem = HistoryItem(historyParts, isUser)
                    history.add(historyItem)
                }
            }

            GeminiX.initChat(
                {
                    call.resolve()
                },
                { error ->
                    call.reject(error)
                }, history)
        } catch (e: Exception) {
            call.reject(e.message)
        }
    }

    @PluginMethod
    fun sendChatMessage(call: PluginCall) {
        try {
            var streamResponse = false
            var images = JSONArray()

            val inputText = call.getString("inputText")!!
            if(call.hasOption("options")){
                val opts = call.getObject("options")
                if(opts.has("streamResponse")){
                    streamResponse = opts.getBoolean("streamResponse")
                }
                if(opts.has("images")){
                    images = opts.getJSONArray("images")
                }
            }

            val modelImages:List<Image> = GeminiX.getModelImages(images, this.context)


            GeminiX.sendChatMessage(
                { response, isComplete ->
                    val result = JSObject()
                    result.put("response", response)
                    if(isComplete){
                        call.resolve(result)
                    }else{
                        result.put("isChat", true)
                        bridge.triggerWindowJSEvent(GeminiXResponseChunkEvent, result.toString())
                    }
                },
                { error ->
                    call.reject(error)
                }, inputText, modelImages, streamResponse)
        } catch (e: Exception) {
            call.reject(e.message)
        }
    }

    @PluginMethod
    fun countChatTokens(call: PluginCall) {
        try {
            var inputText:String? = null
            var images = JSONArray()

            if(call.hasOption("options")){
                val opts = call.getObject("options")
                if(opts.has("images")){
                    images = opts.getJSONArray("images")
                }
                if(opts.has("inputText")){
                    inputText = opts.getString("inputText")
                }
            }

            val modelImages:List<Image> = GeminiX.getModelImages(images, this.context)

            GeminiX.countChatTokens(
                { response ->
                    val result = JSObject()
                    result.put("count", response)
                    result.put("isChat", true)
                    call.resolve(result)
                },
                { error ->
                    call.reject(error)
                }, inputText, modelImages)
        } catch (e: Exception) {
            call.reject(e.message)
        }
    }

    @PluginMethod
    fun getChatHistory(call: PluginCall) {
        try {
            GeminiX.getChatHistory(
                { history ->
                    val result = JSONArray()
                    for (item in history) {
                        val itemJSON = JSONObject()
                        itemJSON.put("isUser", item.isUser)
                        val partsJSON = JSONArray()
                        for (part in item.parts) {
                            val partJSON = JSONObject()
                            when (part) {
                                is TextHistoryPart -> {
                                    partJSON.put("type", "text")
                                    partJSON.put("content", part.content)
                                }
                                is ImageHistoryPart -> {
                                    partJSON.put("type", "image/bitmap")
                                    partJSON.put("content", GeminiX.bitmapToBase64(part.content))
                                }
                                is BlobHistoryPart -> {
                                    partJSON.put("type", part.mimeType)
                                    val contentString = Base64.encodeToString(part.content, Base64.DEFAULT)
                                    partJSON.put("content", contentString)
                                }
                            }
                            partsJSON.put(partJSON)
                        }
                        itemJSON.put("parts", partsJSON)
                        result.put(itemJSON)
                    }
                    call.resolve(JSObject().put("history", result))
                },
                { error ->
                    call.reject(error)
                }
            )
        } catch (e: Exception) {
            call.reject(e.message)
        }
    }
}

File: ios/Sources/GeminiXPlugin/GeminiXPlugin.swift
import Foundation
import Capacitor

/**
 * Please read the Capacitor iOS Plugin Development Guide
 * here: https://capacitorjs.com/docs/plugins/ios
 */
@available(iOS 15.0, *)
@objc(GeminiXPlugin)
public class GeminiXPlugin: CAPPlugin {
    public let identifier = "GeminiXPlugin"
    public let jsName = "GeminiX"
    public let pluginMethods: [CAPPluginMethod] = [
        CAPPluginMethod(name: "initModel", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "sendMessage", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "countTokens", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "initChat", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "sendChatMessage", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "countChatTokens", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "getChatHistory", returnType: CAPPluginReturnPromise)
    ]
    
    let GeminiXResponseChunkEvent = "GeminiXResponseChunk"
    
    /**********************
     *  Plugin API functions
     ************************/
    @objc func initModel(_ call: CAPPluginCall) {
        let params = call.getObject("params") ?? [:];
        GeminiX.initModel(onSuccess: {
            call.resolve()
        }, onError: { error in
            call.reject(error)
        }, params:params)
    }
    
    @objc func sendMessage(_ call: CAPPluginCall) {
        let inputText = call.getString("inputText") ?? ""
        let options = call.getObject("options") ?? [:]
        let streamResponse = call.getBool("streamResponse") ?? false
        let modelImages:[ImageDataWithType] = GeminiX.getImagesFromOptions(options: options)
        
        GeminiX.sendMessage(onSuccess: { response, isComplete, success in
            let result = [
                "response": response,
                "isChat": false
            ]
            if(isComplete){
                call.resolve(result)
            }else{
                let resultString = (try? JSONSerialization.data(withJSONObject: result, options: []))?.base64EncodedString()
                self.bridge?.triggerWindowJSEvent(eventName: self.GeminiXResponseChunkEvent, data: resultString!)
            }
            success()
        }, onError: { error in
            call.reject(error)
        }, inputText: inputText, images: modelImages, streamResponse: streamResponse)
    }
    
    @objc func countTokens(_ call: CAPPluginCall) {
        let inputText = call.getString("inputText") ?? ""
        let options = call.getObject("options") ?? [:]
        let modelImages:[ImageDataWithType] = GeminiX.getImagesFromOptions(options: options)
        
        GeminiX.countTokens(onSuccess: { response in
            let result = [
                "count": response,
                "isChat": false
            ]
            call.resolve(result)
        }, onError: { error in
            call.reject(error)
        }, inputText: inputText, images: modelImages)
    }

    @objc func initChat(_ call: CAPPluginCall) {
        do{
            var history: [HistoryItem] = []
            if(call.hasOption("chatHistory")){
                let history_object = call.getObject("chatHistory") ?? [:]
                let history_json = try JSONSerialization.data(withJSONObject: history_object, options: [])
                if let history_array = try JSONSerialization.jsonObject(with: history_json, options: []) as? [[String: Any]] {
                    for item in history_array {
                        if let isUser = item["isUser"] as? Bool {
                            if let parts = item["parts"] as? [Any]{
                                var historyParts:[HistoryPart] = []
                                for _part in parts {
                                    if let part = _part as? [String:Any]{
                                        if let type = part["type"] as? String {
                                            if type == "text" {
                                                if let content = part["content"] as? String {
                                                    let textPart = TextHistoryPart(content: content)
                                                    historyParts.append(textPart)
                                                }
                                            }
                                            if type == "image" {
                                                if let uri = part["content"] as? String {
                                                    let image:UIImage? = GeminiX.getImageForUri(imageUri: uri)
                                                    if(image != nil){
                                                        let imagePart = ImageHistoryPart(content: image!)
                                                        historyParts.append(imagePart)
                                                    }else{
                                                        throw GeminiXError.invalidArgument("Image file not found: \(uri)")
                                                    }
                                                }
                                            }
                                        }
                                    }
                                }
                                let historyItem = HistoryItem(parts:historyParts, isUser:isUser)
                                history.append(historyItem)
                            }
                        }
                    }
                }
            }
            
            
            GeminiX.initChat(onSuccess: {
                call.resolve()
            }, onError: { error in
                call.reject(error)
            }, history: history)
        }catch{
            call.reject(error.localizedDescription)
        }
    }
    
    @objc func sendChatMessage(_ call: CAPPluginCall) {
        let inputText = call.getString("inputText") ?? ""
        let options = call.getObject("options") ?? [:]
        let streamResponse = call.getBool("streamResponse") ?? false
        let modelImages:[ImageDataWithType] = GeminiX.getImagesFromOptions(options: options)
        
        GeminiX.sendChatMessage(onSuccess: { response, isComplete, success in
            let result = [
                "response": response,
                "isChat": true
            ]
            if(isComplete){
                call.resolve(result)
            }else{
                let resultString = (try? JSONSerialization.data(withJSONObject: result, options: []))?.base64EncodedString()
                self.bridge?.triggerWindowJSEvent(eventName: self.GeminiXResponseChunkEvent, data: resultString!)
            }
            success()
        }, onError: { error in
            call.reject(error)
        }, inputText: inputText, images: modelImages, streamResponse: streamResponse)
    }
    
    @objc func countChatTokens(_ call: CAPPluginCall) {
        let inputText = call.getString("inputText") ?? ""
        let options = call.getObject("options") ?? [:]
        let modelImages:[ImageDataWithType] = GeminiX.getImagesFromOptions(options: options)
        
        GeminiX.countChatTokens(onSuccess: { response in
            let result = [
                "count": response,
                "isChat": true
            ]
            call.resolve(result)
        }, onError: { error in
            call.reject(error)
        }, inputText: inputText, images: modelImages)
    }
    
    @objc func getChatHistory(_ call: CAPPluginCall){
        GeminiX.getChatHistory(onSuccess: { history in
            var historyArray:[[String:Any]] = []
            for item in history {
                var itemParts:[[String:Any]] = []
                for part in item.parts {
                    var partDict:[String:Any] = [:]
                    if let textPart = part as? TextHistoryPart {
                        partDict["type"] = "text"
                        partDict["content"] = textPart.content
                    }
                    if let blobPart = part as? BlobHistoryPart {
                        partDict["type"] = blobPart.mimeType
                        if let content = blobPart.content as? Data{
                            partDict["content"] = content.base64EncodedString()
                        }
                    }
                    itemParts.append(partDict)
                }
                let itemDict:[String:Any] = [
                    "isUser": item.isUser,
                    "parts": itemParts
                ]
                historyArray.append(itemDict)
            }
            var result = JSObject()
            result["history"] = historyArray
            call.resolve(result)
        }, onError: { error in
            call.reject(error)
        })
    }
}
